cf-custom-resources/lib/dns-cert-validator.js

```diff
@@ -0,0 +1,318 @@
+        } else {
+            // Exponential backoff with jitter based on 200ms base
+            // component of backoff fixed to ensure minimum total wait time on
+            // slow targets.
+            const base = Math.pow(2, attempt);
+            await sleep(random() * base * 50 + base * 150);
+        }
+    }
+    if (!record) {
+        throw new Error(`DescribeCertificate did not contain DomainValidationOptions after ${maxAttempts} tries.`)
+    }
+
+    console.log(`Creating DNS record into zone ${hostedZoneId}: ${record.Name} ${record.Type} ${record.Value}`);
+    const changeBatch = await updateRecords(route53, hostedZoneId, 'UPSERT', record.Name, record.Type, record.Value);
+    await waitForRecordChange(route53, changeBatch.ChangeInfo.Id)
+
+    await acm.waitFor('certificateValidated', {
+        // Wait up to 9 minutes and 30 seconds
+        $waiter: {
+            delay: 30,
+            maxAttempts: 19
+        },
+        CertificateArn: reqCertResponse.CertificateArn
+    }).promise();
+
+    return reqCertResponse.CertificateArn;
+};
+
+/**
+ * Deletes a certificate from AWS Certificate Manager (ACM) by its ARN.
+ * If the certificate does not exist, the function will return normally.
+ *
+ * @param {string} arn The certificate ARN
+ */
+const deleteCertificate = async function (arn, region, hostedZoneId) {
+    const [acm, route53] = clients(region)
+    try {
+        console.log(`Waiting for certificate ${arn} to become unused`);
+
+        let inUseByResources;
+        let dnsValidationRecord;
+
+        for (let attempt = 0; attempt < maxAttempts; attempt++) {
+            const {
+                Certificate
+            } = await acm.describeCertificate({
+                CertificateArn: arn
+            }).promise();
+
+            inUseByResources = Certificate.InUseBy || [];
+            dnsValidationRecord = Certificate.DomainValidationOptions || [];
+            if (inUseByResources.length) {
+                // Exponential backoff with jitter based on 200ms base
+                // component of backoff fixed to ensure minimum total wait time on
+                // slow targets.
+                const base = Math.pow(2, attempt);
+                await sleep(random() * base * 50 + base * 150);
+            } else {
+                break
+            }
+        }
+
+        if (inUseByResources.length) {
+            throw new Error(`Certificate still in use after checking for ${maxAttempts} attempts.`)
+        }
+
+        // Fetch the DNS Validation Record and delete it
+        if (dnsValidationRecord.length > 0 && dnsValidationRecord[0].ResourceRecord) {
+            const record = dnsValidationRecord[0].ResourceRecord;
+
+            // Delete this record now that it's not needed
+            const changeBatch = await updateRecords(route53, hostedZoneId, 'DELETE', record.Name, record.Type, record.Value);
+            await waitForRecordChange(route53, changeBatch.ChangeInfo.Id)
```

Comment:
We moved the logic here right? also we're guaranteed right that the certificate will have only one CNAME record and nothing else since we always grab the first element?

Reply:
You can't say that in general, but since we're deleting an ACM cert that was generated by us, and that's guranteed only to have DNS validation options - we can make that claim.